File: contact_custom/static/src/js/custom_contact.js
odoo.define('custom_contact', function (require) {
    "use strict";

    var core = require('web.core');
    var ContactForm = require('contacts.form');

    var CustomContactForm = ContactForm.extend({
        /**
         * Ajoute des événements aux champs CNI et NUI.
         */
        init: function () {
            this._super.apply(this, arguments);

            this.$el.find('[name="cni"]').on('change', this._onCniChange.bind(this));
            this.$el.find('[name="niu"]').on('change', this._onNiuChange.bind(this));
        },

        /**
         * Vérifie le format du numéro CNI.
         */
        _onCniChange: function () {
            var cni = this.$el.find('[name="cni"]').val();
            if (cni.length != 11) {
                this.do_warn(_("Le numéro CNI doit comporter 11 caractères."));
            }
        },

        /**
         * Vérifie le format du numéro NUI.
         */
        _onNiuChange: function () {
            var niu = this.$el.find('[name="niu"]').val();
            if (niu.length != 14) {
                this.do_warn(_("Le numéro NUI doit comporter 14 caractères."));
            }
        },
    });

    core.action_registry.add('custom_contact_form', CustomContactForm);

    return CustomContactForm;
});

File: contact_custom/custom_contact.py
from odoo import api, models, fields, _

class CustomContact(models.Model):
    _inherit = 'res.partner'

    @api.constrains('cni', 'niu')
    def _check_format(self):
        for record in self:
            if record.cni and len(record.cni) != 11:
                raise ValidationError(_("Le numéro CNI doit comporter 11 caractères."))
            if record.niu and len(record.niu) != 14:
                raise ValidationError(_("Le numéro NUI doit comporter 14 caractères."))
